Sources/ForageSDK/Foundation/Network/Collector.swift

```diff
@@ -323,6 +337,157 @@ class BasisTheoryWrapper: VaultCollector {
     }
 }
 
+// Wrapper class for Forage internal vault
+class ForageVaultWrapper: VaultCollector {
+    var customHeaders: [String: String] = [:]
+    let textElement: UITextField
+    
+    private let forageVaultConfig: ForageVaultConfig
+    private let logger: ForageLogger?
+    
+    init(textElement: UITextField, forageVaultConfig: ForageVaultConfig, logger: ForageLogger? = DatadogLogger(ForageLoggerConfig(prefix: "Rosetta"))) {
+        self.textElement = textElement
+        self.customHeaders = [:]
+        self.forageVaultConfig = forageVaultConfig
+        self.logger = logger
+    }
+    
+    func setCustomHeaders(headers: [String : String], xKey: [String : String]) {
+        customHeaders = headers
+    }
+    
+    func sendData<T: Decodable>(path: String, vaultAction: VaultAction, extraData: [String : Any], completion: @escaping (T?, ForageError?) -> Void) {
+        let semaphore = DispatchSemaphore(value: 0)
+        var pin = ""
+        
+        DispatchQueue.main.async {
+            guard let textElementValue = self.textElement.text else {
+                return completion(nil, CommonErrors.INCOMPLETE_PIN_ERROR)
+            }
+            pin = textElementValue
+            semaphore.signal()
+        }
+
+        // waiting for the value of the text field to be read from the main queue and saved in `pin`
+        semaphore.wait()
```

Comment:
why do we need a semaphore here?

Reply:
My experience with this is admittedly limited, but my understanding of what's going on here is this `sendData` method is likely running on a background thread, but reading the value from `UITextField` has to happen on the main thread. Because of that, I'm using `DispatchQueue.main.async { }` to read the value on the main thread, but since these processes are happening in parallel, we can't guarantee we'll read the PIN value before the rest of this method gets called, thus we may not actually submit the PIN if the request happens on the background thread before the main thread updates the pin variable with the value from `UITextField`.

To get around this, it seems like we could either use a completion/callback on the `DispatchQueue.main.async` task (which could be annoying because the rest of the code would have to be nested in that callback, or split into a separate method), or we can take this semaphore approach, which, based on my understanding, suspends running this background thread at the `.wait` call until the main thread calls `.signal` on the semaphore, which only happens once it's successfully read the `UITextField`'s value.

There might be a way easier way to do this that I'm just not aware of since my experience with UIKit/Swift concurrency stuff isn't that extensive 😅

Reply:
refactored to do this in a clearer way here: [8c9c136](https://github.com/teamforage/forage-ios-sdk/pull/169/commits/8c9c136cddfd37aee929d9dcce325cb76000b96c)